File: java-spring/src/main/java/org/burgas/javaspring/service/IdentityDetailsService.java
package org.burgas.javaspring.service;

import lombok.RequiredArgsConstructor;
import org.burgas.javaspring.entity.identity.IdentityDetails;
import org.burgas.javaspring.mapper.IdentityMapper;
import org.jspecify.annotations.NonNull;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Propagation;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true, propagation = Propagation.NOT_SUPPORTED)
public class IdentityDetailsService implements UserDetailsService {

    private final IdentityMapper identityMapper;

    @Override
    public @NonNull UserDetails loadUserByUsername(@NonNull String username) throws UsernameNotFoundException {
        return new IdentityDetails(
                this.identityMapper.identityRepository.findIdentityByEmail(username)
                        .orElseThrow(() -> new IllegalArgumentException("Identity not found for authentication"))
        );
    }
}

File: kotlin-spring/src/main/kotlin/org/burgas/kotlinspring/service/IdentityDetailsService.kt
package org.burgas.kotlinspring.service

import org.burgas.kotlinspring.entity.identity.IdentityDetails
import org.burgas.kotlinspring.mapper.IdentityMapper
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Propagation
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true, propagation = Propagation.NOT_SUPPORTED)
class IdentityDetailsService : UserDetailsService {

    private final val identityMapper: IdentityMapper

    constructor(identityMapper: IdentityMapper) {
        this.identityMapper = identityMapper
    }

    override fun loadUserByUsername(username: String): UserDetails {
        return IdentityDetails(
            this.identityMapper.identityRepository.findIdentityByEmail(username)
                .orElseThrow { throw IllegalArgumentException("Identity not found") }
        )
    }
}